_posts/2018-06-18-configuration.md

```diff
@@ -0,0 +1,162 @@
+      username: System.get_env("EXAMPLE_APP_USERNAME")
+    ]
+    
+    Keyword.merge(opts, system_opts)
+  end
+end
+```
+
+Here `Repo.init/2`, called on start, is used to update our configuration with values from the current system environment without the need to recompile anything.
+
+Now we could generate our build artifacts locally and run them elsewhere.
+How would the aforementioned scenario with `EXAMPLE_APP_HOSTNAME` changes play out?
+An application restart would pull the latest value, no compilation necessary.
+
+Let's update our illustration from before to reflect this new approach:
+
+![elixir-config](https://user-images.githubusercontent.com/73386/41503027-d8b8ecc8-7185-11e8-8284-73d417fea6dc.png)
+
+Our runtime environments have changed but our application's configuration and build artifacts have not, nor should they need to.
+
+We've managed to decouple our code from our runtime configuration with the added bonus of a configuration that is explicit and lives alongside the code.
+
+### Bringing it together
+
+In our last example we see the benefit to separating our configuration into two distinct parts.
+An easy way to avoid the confusion and pitfalls of configuration is to remember this simple rule: `System.get_env/1` should never be used to populate our application's configuration, the values defined in `config.exs`.
+
+Worried about what that means for local development and testing?
+There's no need to fret!
+We can marry these two configuration types to keep things simple and easy for local development.
+
+Let's update our `Repo.init/2` function to reject any values that resolve to `nil` at run time, failing back to the application configuration that's been provided via `opts` (the values set in `config.exs`, `dev.exs`, and `test.exs`).
+
+```elixir
+defmodule ExampleApp.Repo do
+  def init(_, opts) do
+    {:ok, build_opts(opts)}
+  end
+
+  defp build_opts(opts) do
+    system_opts = [
+      database: System.get_env("EXAMPLE_APP_DATABASE"),
+      hostname: System.get_env("EXAMPLE_APP_HOSTNAME"),
+      password: System.get_env("EXAMPLE_APP_PASSWORD"),
+      username: System.get_env("EXAMPLE_APP_USERNAME")
+    ]
+
+    system_opts
+    |> remove_empty_opts()
+    |> merge_opts(opts)
+  end
+
+  defp merge_opts(system_opts, opts) do
+    Keyword.merge(opts, system_opts)
+  end
+
+  defp remove_empty_opts(system_opts) do
+    Enum.reject(system_opts, fn {_k, value} -> is_nil(value) end)
+  end
+end
+```
+
+When our application starts it will attempt to retrieve those system variables, removing the `nil` values, and finally merging the options defined by our application configuration with the runtime configuration, giving precedence to the runtime options.
+
+Now we can use the `dev.exs` and `test.exs` files we're so accustomed to while also ensuring our final build artifact will be correctly set up, thus making configuration of deployments a breeze.
+
+What do you think of this approach?
+We'd love to hear your thoughts!
```

Comment:
Maybe suggest where those thoughts are best directed-- issues, twitter, other...?

Reply:
We need to figure this out TBH.